libraries/ms-common/src/stm32f0xx/interrupt.c

```diff
@@ -0,0 +1,200 @@
+
+// Enables the nested interrupt vector controller for a given EXTI line
+StatusCode prv_nvic_enable(uint8_t irq_channel, InterruptPriority priority) {
+  if (priority >= NUM_INTERRUPT_PRIORITY || irq_channel >= NUM_INTERRUPT_CHANNELS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  NVIC_InitTypeDef init_struct = { .NVIC_IRQChannel = irq_channel,
+                                  .NVIC_IRQChannelPriority = priority,
+                                  .NVIC_IRQChannelCmd = ENABLE };
+
+  NVIC_Init(&init_struct);
+
+  return STATUS_CODE_OK;
+}
+
+#define NUM_INTERRUPT_LINES 32
+
+// Enables the interrupt line with the given settings.
+StatusCode prv_exti_enable(uint8_t line, InterruptSettings *settings) {
+  if (line >= NUM_INTERRUPT_LINES || settings->type >= NUM_INTERRUPT_TYPE ||
+      settings->edge >= NUM_INTERRUPT_EDGE) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  EXTI_InitTypeDef init_struct = { .EXTI_Line = 0x01 << line,
+                                  .EXTI_Mode = 0x04 * settings->type,
+                                  .EXTI_Trigger = 0x08 + 0x04 * settings->edge,
+                                  .EXTI_LineCmd = ENABLE };
+  EXTI_Init(&init_struct);
+
+  return STATUS_CODE_OK;
+}
+
+// Trigger a software interrupt on an EXTI line if that line supports this functionality.
+StatusCode prv_exti_trigger_interrupt(uint8_t line) {
+  // It is not possible to manually trigger interrupts for lines 18 and > 22.
+  if (line == 18 || line > 22) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  EXTI_GenerateSWInterrupt(0x01 << line);
+  return STATUS_CODE_OK;
+}
+
+// Gets the value of the pending flag for a given interrupt. A pending_bit that gets set to 1
+// indicates an interrupt is waiting.
+StatusCode prv_exti_get_pending(uint8_t line, uint8_t *pending_bit) {
+  if (line >= NUM_INTERRUPT_LINES) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  *pending_bit = EXTI_GetITStatus(0x01 << line);
+  return STATUS_CODE_OK;
+}
+
+// Clears the pending flag for the interrupt.
+StatusCode prv_exti_clear_pending(uint8_t line) {
+  if (line >= NUM_INTERRUPT_LINES) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  EXTI_ClearITPendingBit(0x01 << line);
+  return STATUS_CODE_OK;
+}
+
+// GPIO INTERRUPTS
+
+static interrupt_callback s_interrupt_gpio_callback[NUM_GPIO_PINS];
+
+// Helper function to enable gpio interrupts.
+StatusCode prv_gpio_interrupt_enable(GPIOAddress *address, InterruptSettings *settings,
+                                     interrupt_callback callback, uint8_t irq_channel) {
+  SYSCFG_EXTILineConfig(address->port, address->pin);
+  status_ok_or_return(prv_exti_enable(address->pin, settings));
+  s_interrupt_gpio_callback[address->pin] = callback;
+  status_ok_or_return(prv_nvic_enable(irq_channel, settings->priority));
+  return STATUS_CODE_OK;
+}
+
+StatusCode interrupt_gpio_register(GPIOAddress *address, InterruptSettings *settings,
```

Comment:
The only problem I have with this is a ton of internal, reusable code that should be private is in this module. I really don't want to rewrite parts of this in every interrupt file (besides IDs and things should be tracked globally for x86) and I don't really want to expose the registration functions as it makes the API really bad. Especially because it shouldn't be called from any modules other than other ms-common files. I would only be in favour of this if we moved the core interrupt code to its own library (under the STM PERIPH LIB and start a new lib for x86 which can be included in the implementation code of interrupts for each module). The same this is going to happen with repeated code in timers which should probably get broken into hardware timers and software timers.